mlflow/utils/environment.py

```diff
@@ -464,7 +464,8 @@ def _generate_mlflow_version_pinning():
     mlflow_version = Version(VERSION)
     current_major_version = mlflow_version.major
     current_minor_version = mlflow_version.minor
-    return f"mlflow=={current_major_version}.{current_minor_version}"
+    current_micro_version = mlflow_version.micro
+    return f"mlflow=={current_major_version}.{current_minor_version}.*"```
```

Comment:
To install the latest version, use `*`.

Reply:
Mh.. interesting idea... 
I guess we can do that, but must be really careful not to introduce introducing incompatibilities between micro versions. 

Hardcoding the micro version might be more stable, but removes a bit of flexibility. 
It's up to you but I rather would hardcode it, it's very easy to mess up at some point without realizing

Reply:
> Hardcoding the micro version might be more stable

This also has a risk. Let's there is a version (1.2.3) that contains a bug in the serving logic. If your model is logged with `1.2.3`, you can't server your model. If your model is logged with `1.2.*`, you can serve the model after 1.2.4 is released.

Reply:
Here is also a scenario where this could be a foot gun: 

1. User has `mlflow==2.4.0`

2. User uses mlflow-serve
The deployed model now uses `mlflow==2.4.LATEST`


If there are any changes in API's in the mlflow tracking/server functionalities, things could very easily break here.


So let's really just hardcode the micro version. Users will not expect the that the version of mlflow of the serving venv and the "master venv" may differ and may cause many other troubles down the road

Reply:
>  If your model is logged with 1.2.3, you can't server your model. If your model is logged with 1.2.*, you can serve the model after 1.2.4 is released.

Yeah using `*` gives you the ability to make `sneaky patches` and fix stuff.
But you can also put in `sneaky bugs` without realizing :P

Reply:
> Yeah using * gives you the ability to make sneaky patches and fix stuff.
But you can also put in sneaky bugs without realizing :P

100% agree. There is a tradeoff.

Reply:
Exactly. I am leaning a bit more toward hard coding.

Sneakily breaking somebody's models can be worse than sneakily fixing them.


------


Imagine 20k zipped models sitting around somewhere and suddenly breaking because of a mlflow update. `sneaky bug`

----
But on the other hand, mlflow could accidentally release a bug one day, breaking 20k models first. They could be `sneaky patched` by releasing another mlflow version. 


But this also means all micro versions of mlflow must essentially be backward compatible. 
I.e. `mlflow==A.B.C` must work with any other `mlflow==A.B.Z`  where `C` and `Z` could be any number. 


To support this you must be super careful and theoretically also test it in CI. 
It's up to Mlflow team if you guys want to bit the bullet of backwards compatibility :P 
If done perfectly, it would be better. 
But it's really tough, especially for a complex project like mlflow and maybe not worth the effort.

Reply:
Discussed offline, I think we should use strict pinning to micro version A.B.C, such as 2.4.1

Reply:
We discussed this in the team. Let's log the micro version.